pkg/gardenlet/apis/config/validation/validation.go

```diff
@@ -96,9 +96,10 @@ func ValidateGardenletConfiguration(cfg *config.GardenletConfiguration, fldPath
 		}
 
 		if handler.SNI != nil && handler.SNI.Ingress != nil && handler.SNI.Ingress.ServiceExternalIP != nil {
-			if !cfg.FeatureGates[string(features.APIServerSNI)] {
+			if apiServerSNIEnabled, ok := cfg.FeatureGates[string(features.APIServerSNI)]; ok && !apiServerSNIEnabled {
```

Comment:
I think, this feature gate check is actually conceptually wrong at this place and should rather be removed.

Why? The validation code is called in two places: in g-api (for validation of `ManagedSeed`s for example) and in gardenlet (on startup). But actually, only the second place will know exactly, which feature gates are enabled by default and so on (version skew between g-api and gardenlet / custom image / etc...). So the first check cannot foresee, how the feature gate enablement will actually look like on the gardenlet side at the end, so it also shouldn't try to predict / validate it. Therefore it should be removed from the common validation code.

If it's important to prevent such misconfiguration, I would
- add a dedicated validation func, that can be called by gardenlet on startup (i.e. when we know exactly what feature gates are enabled)
- add a simple check where the stuff is used and then fail or just log an error

Or if it's not too important (which I assume, is the case), we can silently ignore the specific field.

Reply:
Why not already fail at this stage (`ManagedSeed` validation) when the `APIServerSNI` feature flag is set to `false`? I mean that is an invalid input in combination with an ExposureClass handler that have `serviceExternalIP != nil`, even if the final Gardenlet configuration might look different. So we prevent that no semantic invalid Gardenlet configuration is saved in the `ManagedSeed` resource. So I think failing is the correct reaction, isn't it?

For the second stage (config validation on Gardenlet startup) we would again check the same, but this time maybe with changed configuration inputs, but the validation logic need to be the same (`.featureGates["APIServerSNI"]=false && .exposureClassHandlers[].sni.ingress.serviceExternalIP != nil` is invalid).

In my opinion it would be good to have the validation in both stages as we then ensure that no invalid configuration is stored in the `ManagedSeed` resource and fail at design time already and not only at runtime later.

> Or if it's not too important (which I assume, is the case), we can silently ignore the specific field.

Hmm you mean if `.featureGates["APIServerSNI"]=false && .exposureClassHandlers[].sni.ingress.serviceExternalIP != nil != nil` we should just log or ignore it? I would not vote for doing that as the operator expects if an ip is specified via `serviceExternalIP` that it will get assigned to the load balancer service of ExposureClass handler and this is only possible if the feature `APIServerSNI` is inactive. Otherwise it does not make any sense (non sni means every control plane get its own `kube-apiserver` lb service and we cannot assign a single ip to many lb services).

Reply:
Ok, we had a short chat about this and it turns out, I was looking at it the wrong way this morning.
The check is a best-effort validation to prevent operators from actively disabling the feature gate and it's not about making sure that the feature gate is enabled (which we can't know during validation in the apiserver).
As the feature gate is already beta in the version that rolls out the `ExposureClass` feature, it is enabled by default and we only want to prevent operators from actively disabling it.

However, the validation should not fail, if the feature gate is not actively enabled in the config. This was not the case before and is now fixed with this PR.